src/client/gameannouncer.py

```diff
@@ -1,61 +1,60 @@
-from PyQt5.QtCore import QTimer
 from model.game import GameState
-
-from fa import maps
-
+from chat.friendtracker import build_friends_tracker
 
 class GameAnnouncer:
-    ANNOUNCE_DELAY_SECS = 35
 
-    def __init__(self, gameset, me, colors, client):
-        self._gameset = gameset
+    def __init__(self, playerset, me, colors, client):
         self._me = me
         self._colors = colors
         self._client = client
 
-        self._gameset.newLobby.connect(self._announce_hosting)
-        self._gameset.newLiveReplay.connect(self._announce_replay)
+        self._friends_event_tracker = build_friends_tracker(me, playerset)
+        self._friends_event_tracker.friendEvent.connect(self._friend_event)
 
         self.announce_games = True
         self.announce_replays = True
-        self._delayed_host_list = []
-
-    def _is_friend_host(self, game):
-        return (game.host_player is not None
-                and self._me.isFriend(game.host_player.id))
-
-    def _announce_hosting(self, game):
-        if not self._is_friend_host(game) or not self.announce_games:
-            return
-        announce_delay = QTimer()
-        announce_delay.setSingleShot(True)
-        announce_delay.setInterval(self.ANNOUNCE_DELAY_SECS * 1000)
-        announce_delay.timeout.connect(self._delayed_announce_hosting)
-        announce_delay.start()
-        self._delayed_host_list.append((announce_delay, game))
+        self._delayed_event_list = []
+        self.delay_friend_events = True
 
-    def _delayed_announce_hosting(self):
-        timer, game = self._delayed_host_list.pop(0)
+    def _friend_event(self, player, event):
+        if self.delay_friend_events:
+            self._delayed_event_list.append((player, event))
+        else:
+            self._friend_announce(player, event)
 
-        if (not self._is_friend_host(game) or
-           not self.announce_games or
-           game.state != GameState.OPEN):
-            return
-        self._announce(game, "hosting")
+    def delayed_friend_events(self):
+        self.delay_friend_events = False
+        while len(self._delayed_event_list) > 0:
+            player, event = self._delayed_event_list.pop(0)
+            self._friend_announce(player, event)
 
-    def _announce_replay(self, game):
-        if not self._is_friend_host(game) or not self.announce_replays:
+    def _friend_announce(self, player, event):
```

Comment:
I don't really like how we don't use the event type here at all - it means that the logic for figuring out what event happened to a player is in two places, here and in friend tracker. If we change something, we'll have to change code in two places. Can we instead use the event type, and only do some basic sanity checks (like if the game is open)?

Reply:
We can do that ...